### PR TITLE
lab 02 updated, issues raised by bot fixed

### LINKED ISSUE #441: Lab 02 Bubblesort and Sorting
Create an executable go program in directory 02_bubble/marchitect
Write a function that returns a sorted copy of int slice s using bubble sort
Bonus points: implement Insertion sort
Extra bonus points: implement an O(n log(n)) sorting algorithm

### PR DESCRIPTION
**Pull request by Marchitect**
Fixes #441

Review of colleague's PR #438 

Changes proposed in this PR:
- added sorting functions and test cases, fixed issues raised by bot

### CODE REVIEW

File: 02_bubble/marchitect/main.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+			// if the current element is greater than the next element, swap them
+			if s[i-1] > s[i] {
+				// swap values - Go tuple assignment
+				s[i], s[i-1] = s[i-1], s[i]
+				// set swap back to true
+				swapped = true
+			}
+		}
+	}
+	return s
```
Comment (davidbroughsmyth) [inline 02_bubble/marchitect/main.go]: The lab asks for returning a sorted copy of int slice s. Returning s is not a copy.
You can make a copy by making a slice then copy:
```
copy := make([]int, len(s))
copy(copy, s)
```

**Review by davidbroughsmyth — CHANGES_REQUESTED** [review]
Need to change sort functions to returns a sorted copy of int slice s

**Comment by davidbroughsmyth** [comment]
You will also need to review a colleague's PR.

File: 02_bubble/marchitect/main.go (hunk 1 above)
Comment (Marchitect) [reply 02_bubble/marchitect/main.go]: Thank you for the feedback, will improve the code accordingly. And yes I aim to spare some time to review more PR too

Comment (Marchitect) [reply 02_bubble/marchitect/main.go]: > You will also need to review a colleague's PR.

Kindly please help me check if the updated commit is good to fix the issue :) Thank you David.

Comment (davidbroughsmyth) [reply 02_bubble/marchitect/main.go]: Nicely done. Looks good to me.

**Comment by davidbroughsmyth** [comment]
And I like how you made use of `if !reflect.DeepEqual(actual, expected) {`in your tests.

File: 02_bubble/marchitect/main.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"io"
+	"math/big"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func bubble(s []int) []int {
+	// n is the number of items in the int array
```
Comment (juliaogris) [inline 02_bubble/marchitect/main.go:14]: don't comment the obvious (below as well).

File: 02_bubble/marchitect/main.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		return result
+	}
+	// set swapped to true
+	swapped := true
+	// in Go, use keyword for to act as a while loop
+	for swapped {
+		// set swapped to false
+		swapped = false
+		// iterate through all of the elements in our list
+		for i := 1; i < n; i++ {
```
Comment (juliaogris) [inline 02_bubble/marchitect/main.go:30]: The inner loop usually goes from `n` down. Reason: after the first outer iteration the largest element has "bubbled up" to the slice position with the higher index.
Our reference solution looks like this - feel free to use, adjust or ignore:
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			// If no elements were swapped, the slice is sorted so we can stop here
			return
		}
	}
```

File: 02_bubble/marchitect/main.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+func bubble(s []int) []int {
+	// n is the number of items in the int array
+	n := len(s)
+	// make a copy of the original int array s
+	result := make([]int, n)
+	copy(result, s)
+	// maybe necessary
+	if n < 2 {
+		return result
+	}
```
Comment (juliaogris) [inline 02_bubble/marchitect/main.go:22]: nope ;) not necessary.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Looks pretty good.
I think you have gone overboard a bit with the comments, which IMO hinders readability.

**Pull request closed without merging**